sql/src/main/java/io/crate/operation/collect/collectors/CompositeCollector.java

```diff
@@ -80,141 +73,102 @@
 
-        private final RowReceiver delegate;
-        private final AtomicInteger activeUpstreams = new AtomicInteger(0);
-        private final List<RepeatHandle> repeatHandles = new ArrayList<>();
-        private Throwable failure = null;
-        private boolean prepared = false;
-        private CompletionListener listener = CompletionListener.NO_OP;
-
-        MultiRowReceiver(RowReceiver delegate) {
-            this.delegate = delegate;
-        }
-
-        private void countdown() {
-            int numUpstreams = activeUpstreams.decrementAndGet();
-            LOGGER.trace("countdown numUpstreams={} failure={}", numUpstreams, failure);
-            if (numUpstreams == 0) {
-                if (failure == null) {
-                    final ImmutableList<RepeatHandle> repeatHandles = ImmutableList.copyOf(this.repeatHandles);
-                    this.repeatHandles.clear();
-                    delegate.finish(new RepeatHandle() {
-                        @Override
-                        public void repeat() {
-                            if (MultiRowReceiver.this.activeUpstreams.compareAndSet(0, repeatHandles.size())) {
-                                for (RepeatHandle repeatHandle : repeatHandles) {
-                                    repeatHandle.repeat();
-                                }
-                            } else {
-                                throw new IllegalStateException("Repeat called without all upstreams being finished");
-                            }
-                        }
-                    });
-                } else {
-                    delegate.fail(failure);
-                }
-            }
-        }
-
-        @Override
-        public RowReceiver newRowReceiver() {
-            int numUpstreams = activeUpstreams.incrementAndGet();
-            LOGGER.trace("newRowReceiver activeUpstreams={}", numUpstreams);
-            return this;
-        }
+        // holds the number of upstream finish calls
+        // this field is set to -1 to act as a kill marker
+        private volatile int numFinishCalls = 0;
 
         @Override
         public Result setNextRow(Row row) {
-            return delegate.setNextRow(row);
+            return downstream.setNextRow(row);
         }
 
         @Override
         public void pauseProcessed(ResumeHandle resumeable) {
-            delegate.pauseProcessed(resumeable);
+            downstream.pauseProcessed(resumeable);
+        }
+
+        public void repeat() {
+            assert numFinishCalls % collectors.size() == 0 : "Repeat called without all upstreams being finished";
+            if (numFinishCalls == -1) return;
+            repeatHandles[0].repeat();
         }
 
         @Override
         public void finish(RepeatHandle repeatable) {
-            this.repeatHandles.add(repeatable);
-            countdown();
-            listener.onSuccess(null);
+            if (numFinishCalls == -1) return;
```

Comment:
might make sense to assign the volatile variable to a local variable to avoid multiple volatile reads